### PR TITLE
UC-50

### PR DESCRIPTION
**Pull request by cpatelss90**

### CODE REVIEW

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    
+	/**
+	 * 
+	 * @param userId, loanRequest(body)
+	 * @return void
+	 * 
+	 */
+    
+    
+	@PostMapping("/loansignup")
```
Comment (Joshua-Podhola) [inline src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:74]: Non-API compliant URI.

**Review by Joshua-Podhola — CHANGES_REQUESTED** [review]
I can't even compile and run this in the current state (see comments), among many issues. Keep in mind that I'm looking at all of this before the back end, but I'm quite concerned.

There's also a complete lack of unit testing, but given the way the code is structured, I can see why.

File: src/main/java/com/ss/ucm/ms/loans/entities/UserLoan.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//	@JoinColumn(name = "loan_id", nullable = false)
+//	private Loan loan;
+//	
+//    
+//	@ManyToOne(fetch = FetchType.LAZY, optional = false)
+//    @JoinColumn(name = "users_id", nullable = false)
+//    private User user;
+    
+    @Id
+    private int loanId;
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/entities/UserLoan.java:46]: No, the UserLoan Table has a composite key of loadID, userID but the eclipse is crying and doesn't run if I don't specify @Id annotation

File: src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Class for composite key
+ * 
+ * 
+ * @author Charvin Patel
+ */
+
+
+@IdClass(LoanIdUserId.class)
+public class LoanIdUserId implements Serializable {
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java:16]: This is need for a composite key for a UserLoan table which is now being used.

File: src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Class for composite key
+ * 
+ * 
+ * @author Charvin Patel
+ */
+
+
+@IdClass(LoanIdUserId.class)
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java:15]: Fixed.

File: src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+public class LoanIdUserId implements Serializable {
+
+	
+	/**
+	 * 
+	 */
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/entities/LoanIdUserId.java:21]: Just a slip up. Fixed.

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	public void signupLoan(@RequestParam int userId,  @RequestBody RequestLoanSignupDto loanRequest) {
+			
+		Loan loan = new Loan();
+		
+		loan.setMaxAmount(loanRequest.getMaxAmount());
+		loan.setName(loanRequest.getName());
+		loan.setInterestRate(loanRequest.getInterestRate());
+				
+		loanAdd.saveLoan(loan);
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:83]: They are not creating any loan types.....I am giving them just 3 options...either "home ","auto","personal" loan in which, the users selects one and get inserted into the database

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+		UserLoan userLoan = new UserLoan();
+		userLoan.setLoanId(loanSearch.LoanId(loanRequest.getName()));
+		userLoan.setUserId(userId);
+		userLoan.setBalance(loanRequest.getBalance());
+		userLoan.setDate(loanRequest.getStartDate().now());
+				
+		userLoanAdd.saveUserLoan(userLoan);
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:91]: Fixed.

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java (hunk 1 above)
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:74]: not sure what API compliant are we even using? Did we discuss about this?

File: src/main/java/com/ss/ucm/ms/loans/entities/Loan.java [hunk 8]
```diff
@@ -13,7 +16,7 @@
     //Data
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
-    private Integer id;
+    private int id;
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/entities/Loan.java:19]: not any reasons

File: pom.xml [hunk 9]
```diff
@@ -1,67 +1,84 @@
...
+<!-- 		<dependency>
+			<groupId>com.ucm</groupId>
+			<artifactId>lib_auth</artifactId>
+			<version>0.0.2-SNAPSHOT</version>
+		</dependency> -->
+		<dependency>
+			<groupId>com.ucm.ms</groupId>
+			<artifactId>ucm-ms-users</artifactId>
+			<version>0.0.1-SNAPSHOT</version>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply pom.xml:71]: Fixed.

File: src/main/java/com/ss/ucm/ms/loans/dto/RequestLoanSignupDto.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	private String name;
+
+	@NotBlank
+	private BigDecimal interest_rate;
+
+	@NotBlank
+	private BigDecimal balance;
+
+	@NotBlank
+	private LocalDate startDate;
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/dto/RequestLoanSignupDto.java:35]: Nope wrong, they are not creating any loan types.....I am giving them just 3 options...either "home ","auto","personal" loan in which, the users selects one  which then needs to get inserted into database in which we obviously need backend, can't just have it done with frontend.

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -2,12 +2,14 @@
...
 spring.datasource.username=sa
-spring.datasource.password=password
+#spring.datasource.password=password
 spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
 spring.h2.console.enabled=true
-spring.h2.console.path=/h2
+#spring.h2.console.path=/h2
 spring.h2.console.settings.trace=false
 spring.h2.console.settings.web-allow-others=false
 spring.jpa.hibernate.ddl-auto=validate
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/resources/application.properties:11]: Not sure what happened there but now fixed.

File: src/main/java/com/ss/ucm/ms/loans/services/LoanSearch.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    
+	/**
+	 * 
+	 * @param name
+	 * @return user_id(int)
+	 * 
+	 */
+    public int LoanId(String name) {
+    	 return loanDAO.getIdByName(name);
+    }
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/services/LoanSearch.java:37]: See my front-end it has static loan information which user will select.....also see my comment about "future improvement"

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+	/**
+	 * 
+	 * @param userId, loanRequest(body)
+	 * @return void
+	 * 
+	 */
+    
+    
+	@PostMapping("/loansignup")
+	public void signupLoan(@RequestParam int userId,  @RequestBody RequestLoanSignupDto loanRequest) {
```
[the remark this answers is not in the document]
Comment (cpatelss90) [reply src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:75]: need to setup a small meeting preferably after monday's demo to understand what you actually did in lib-auth

File: src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java (hunk 1 above)
Comment (Joshua-Podhola) [reply src/main/java/com/ss/ucm/ms/loans/controllers/LoansController.java:74]: It was about Akbar's lecture a few weeks back

More specifically, the URL shouldn't contain verbs and should point to a resource depending on how to reach it. For example, POST .../loan/ would be for posting a new loan. Can just change it later I suppose.

**Review by Joshua-Podhola — APPROVED** [review]
Runs just fine. I'm not sure having static content on the front end and posting a new loan type every time the user signs up for one (and having that information gotten from the front end) is a great idea, but I don't think we can really afford for me to hold up development because of that.

I'll just warn that Patrick might not be very happy about the way it was done.

**Pull request merged**